src/raftstore/store/worker/unsafe_cleanup_range.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use rocksdb::{ReadOptions, SeekKey, Writable, DB};
+
+use util::escape;
+use util::timer::Timer;
+use util::worker::{Runnable, RunnableWithTimer};
+use raftstore::store::{keys, util};
+use storage::{CF_DEFAULT, CF_LOCK, CF_RAFT, CF_WRITE};
+use util::rocksdb::get_cf_handle;
+
+pub const UNSAFE_CLEANUP_INTERVAL: u64 = 5_000; // milliseconds
+
+#[derive(Debug)]
+pub struct TaskQueue {
+    db: Arc<DB>, // use db to store tasks.
+}
+
+impl Clone for TaskQueue {
+    fn clone(&self) -> TaskQueue {
+        TaskQueue {
+            db: Arc::clone(&self.db),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Task {
+    pub cf: String,
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+}
+
+impl Task {
+    fn new(cf: String, start_key: Vec<u8>, end_key: Vec<u8>) -> Task {
+        Task {
+            cf: cf,
+            start_key: start_key,
+            end_key: end_key,
+        }
+    }
+}
+
+pub fn encode_cf(cf: &str) -> u8 {
+    match cf {
+        CF_DEFAULT => 0x01,
+        CF_LOCK => 0x02,
+        CF_WRITE => 0x03,
+        CF_RAFT => 0x04,
+        _ => panic!("unknow column family {}", cf),
+    }
+}
+
+pub fn decode_cf(cf: u8) -> &'static str {
+    match cf {
+        0x01 => CF_DEFAULT,
+        0x02 => CF_LOCK,
+        0x03 => CF_WRITE,
+        0x04 => CF_RAFT,
+        _ => panic!("unkown column family code {}", cf),
+    }
+}
+
+fn encode_task_key(cf: &str, start_key: &[u8]) -> Vec<u8> {
+    let mut vec = keys::encode_unsafe_cleanup_range_key(start_key);
+    vec.push(encode_cf(cf));
+    vec
+}
+
+fn decode_task_key(key: &[u8]) -> (&'static str, Vec<u8>) {
+    let len = key.len();
+    assert!(len >= keys::UNSAFE_CLEANUP_RANGE_MIN_KEY.len() + 1);
+    let cf = decode_cf(key[len - 1]);
+    let start_key = keys::decode_unsafe_cleanup_range_key(&key[..len - 1])
+        .unwrap_or_else(|e| panic!("decode key failed: {:?}", e));
+    (cf, start_key)
+}
+
+impl TaskQueue {
+    pub fn new(db: Arc<DB>) -> TaskQueue {
+        TaskQueue { db: db }
+    }
+
+    pub fn add_task(&self, cf: &str, start_key: &[u8], end_key: &[u8]) -> Result<(), String> {
```

Comment:
Is there a possibility that task key has already existed in the db, in default cf?

Reply:
The `UNSAFE_CLEANUP_RANGE_PREFIX` distinguishes these keys from others